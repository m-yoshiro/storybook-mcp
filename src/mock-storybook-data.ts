import type { Component } from './storybook-api.js';

export const mockComponents: Component[] = [
  {
    id: 'components/mycomponent',
    name: 'Components/MyComponent',
    description: '',
    props: [],
    stories: {
      'Components/MyComponent': {
        id: 'components-mycomponent--simple',
        title: 'Components/MyComponent',
        name: 'Simple',
        importPath: './src/components/MyComponent.stories.jsx',
        kind: 'Components/MyComponent',
        story: 'Simple',
        parameters: {
          __id: 'components-mycomponent--simple',
          docsOnly: false,
          fileName: './src/components/MyComponent.stories.jsx',
        },
      },
    },
  },
  {
    id: 'components/myothercomponent',
    name: 'Components/MyOtherComponent',
    description: '',
    props: [],
    stories: {
      'Components/MyOtherComponent': {
        id: 'components-myothercomponent--simple',
        title: 'Components/MyOtherComponent',
        name: 'Simple',
        importPath: './src/components/MyOtherComponent.stories.jsx',
        kind: 'Example/Button',
        story: 'Simple',
        parameters: {
          __id: 'components-myothercomponent--simple',
          docsOnly: false,
          fileName: './src/components/MyOtherComponent.stories.jsx',
        },
      },
    },
  },
];
