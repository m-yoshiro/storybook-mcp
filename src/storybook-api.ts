import { promises as fs } from 'node:fs';
import path from 'node:path';
import { getConfig } from './config.js';

// Storybook json types
// https://github.com/storybookjs/storybook/blob/next/code/core/src/types/modules/indexer.ts

export interface Parameters {
  __id: string;
    docsOnly: boolean;
    fileName: string;
  // biome-ignore lint/suspicious/noExplicitAny: we can't expect the exact structure of the parameters
  [name: string]: any;
}

export interface ComponentProp {
  name: string;
  type: string;
  // biome-ignore lint/suspicious/noExplicitAny: we can't expect the type of the defaultValue
  defaultValue: any;
  description?: string;
}

export interface StorybookStory {
  id: string;
  title: string;
  name: string;
  importPath: string;
  kind: string;
  story: string;
  parameters: Parameters;
}

export interface ComponentStory {
  name: string;
  parameters: Parameters;
  id?: string;
  title?: string;
  importPath?: string;
  kind?: string;
  story?: string;
}

export interface Component {
  id: string;
  name: string;
  description?: string;
  props: ComponentProp[];
  stories: Record<string, ComponentStory>;
}

export interface StorybookData {
  v: number;
  stories: {
    [key: string]: ComponentStory;
  };
}

export const getComponents = async (storybookStaticDir: string): Promise<Component[]> => {
  try {
    const config = getConfig();

    // Use the provided path or the one from config
    let dirPath = storybookStaticDir || config.storybookStaticDir;

    // Check if the path is relative, convert to absolute if needed
    if (!path.isAbsolute(dirPath)) {
      dirPath = path.resolve(process.cwd(), dirPath);
      console.log('Converted relative path to absolute:', dirPath);
    }

    // Only join 'stories.json' if it's not already included in the path
    let storiesJsonPath = dirPath;
    if (!dirPath.endsWith('stories.json')) {
      storiesJsonPath = path.join(dirPath, 'stories.json');
    }
    console.log('storiesJsonPath:', storiesJsonPath);

    const storiesJsonContent = await fs.readFile(storiesJsonPath, 'utf-8');
    console.log('storiesJsonContent:', storiesJsonContent);
    const storiesData: StorybookData = JSON.parse(storiesJsonContent);
    console.log('storiesData:', storiesData);

    const components: Component[] = [];

    for (const storyId in storiesData.stories) {
      console.log('storyId:', storyId);
      const story = storiesData.stories[storyId];

      if (!story.title) {
        continue; // Skip stories without a title
      }

      // Check if the story already exists in the components array
      let component = components.find(c => c.name === story.title);

      if (!component) {
        component = {
          id: story.title.toLowerCase().replace(/ /g, '-'),
          name: story.title,
          description: '',
          props: [],
          stories: {},
        };
        components.push(component);
      }

      if (component) {
        component.stories[story.title] = {
          name: story.name,
          parameters: story.parameters,
          id: story.id,
          title: story.title,
          importPath: story.importPath,
          kind: story.kind,
          story: story.story,
        };
      }
    }

    return components;
  } catch (error) {
    console.error('Error reading or parsing stories.json:', error);
    return [];
  }
};
