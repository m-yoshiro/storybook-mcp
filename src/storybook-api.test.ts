import { describe, expect, it } from 'bun:test';
import { mockComponents } from './mock-storybook-data.js';
import { getComponents } from './storybook-api.js';

const mockStoryPath = './src/__mocks__/stories.json';

describe('storybook-api', () => {
  it('should return mock components data', async () => {
    const components = await getComponents(mockStoryPath);

    // Verify components are the mock components
    expect(components).toEqual(mockComponents);
  });

  it('should return components with expected structure', async () => {
    const components = await getComponents(mockStoryPath);

    // Check that we got an array
    expect(Array.isArray(components)).toBe(true);

    // Check that at least one component exists
    expect(components.length).toBeGreaterThan(0);

    // Check the structure of the first component
    const component = components[0];
    expect(component).toHaveProperty('name');
    expect(component).toHaveProperty('description');
    expect(component).toHaveProperty('props');
    expect(component).toHaveProperty('stories');

    // Check props array structure
    expect(Array.isArray(component.props)).toBe(true);
    if (component.props.length > 0) {
      const prop = component.props[0];
      expect(prop).toHaveProperty('name');
      expect(prop).toHaveProperty('type');
      expect(prop).toHaveProperty('defaultValue');
    }

    // Check stories object structure
    expect(typeof component.stories).toBe('object');
    if (Object.keys(component.stories).length > 0) {
      const storyKey = Object.keys(component.stories)[0];
      const story = component.stories[storyKey];
      expect(story).toHaveProperty('name');
      expect(story).toHaveProperty('parameters');
    }
  });

  it('should match snapshot for component structure', async () => {
    const components = await getComponents(mockStoryPath);
    expect(components).toMatchSnapshot();
  });

  it('should match snapshot for individual components', async () => {
    const components = await getComponents(mockStoryPath);

    // Test snapshots for each component
    components.forEach((component, index) => {
      expect(component).toMatchSnapshot(`component-${index}-${component.name}`);

      // Test snapshots for component properties
      expect(component.props).toMatchSnapshot(`component-${index}-${component.name}-props`);

      // Test snapshots for component stories
      for (const [storyKey, story] of Object.entries(component.stories)) {
        expect(story).toMatchSnapshot(`component-${index}-${component.name}-story-${storyKey}`);
      }
    });
  });
});
