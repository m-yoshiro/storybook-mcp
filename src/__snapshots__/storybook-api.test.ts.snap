// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`storybook-api should match snapshot for component structure 1`] = `
[
  {
    "description": "",
    "id": "components/mycomponent",
    "name": "Components/MyComponent",
    "props": [],
    "stories": {
      "Components/MyComponent": {
        "id": "components-mycomponent--simple",
        "importPath": "./src/components/MyComponent.stories.jsx",
        "kind": "Components/MyComponent",
        "name": "Simple",
        "parameters": {
          "__id": "components-mycomponent--simple",
          "docsOnly": false,
          "fileName": "./src/components/MyComponent.stories.jsx",
        },
        "story": "Simple",
        "title": "Components/MyComponent",
      },
    },
  },
  {
    "description": "",
    "id": "components/myothercomponent",
    "name": "Components/MyOtherComponent",
    "props": [],
    "stories": {
      "Components/MyOtherComponent": {
        "id": "components-myothercomponent--simple",
        "importPath": "./src/components/MyOtherComponent.stories.jsx",
        "kind": "Example/Button",
        "name": "Simple",
        "parameters": {
          "__id": "components-myothercomponent--simple",
          "docsOnly": false,
          "fileName": "./src/components/MyOtherComponent.stories.jsx",
        },
        "story": "Simple",
        "title": "Components/MyOtherComponent",
      },
    },
  },
]
`;

exports[`storybook-api should match snapshot for individual components: component-0-Components/MyComponent 1`] = `
{
  "description": "",
  "id": "components/mycomponent",
  "name": "Components/MyComponent",
  "props": [],
  "stories": {
    "Components/MyComponent": {
      "id": "components-mycomponent--simple",
      "importPath": "./src/components/MyComponent.stories.jsx",
      "kind": "Components/MyComponent",
      "name": "Simple",
      "parameters": {
        "__id": "components-mycomponent--simple",
        "docsOnly": false,
        "fileName": "./src/components/MyComponent.stories.jsx",
      },
      "story": "Simple",
      "title": "Components/MyComponent",
    },
  },
}
`;

exports[`storybook-api should match snapshot for individual components: component-0-Components/MyComponent-props 1`] = `[]`;

exports[`storybook-api should match snapshot for individual components: component-0-Components/MyComponent-story-Components/MyComponent 1`] = `
{
  "id": "components-mycomponent--simple",
  "importPath": "./src/components/MyComponent.stories.jsx",
  "kind": "Components/MyComponent",
  "name": "Simple",
  "parameters": {
    "__id": "components-mycomponent--simple",
    "docsOnly": false,
    "fileName": "./src/components/MyComponent.stories.jsx",
  },
  "story": "Simple",
  "title": "Components/MyComponent",
}
`;

exports[`storybook-api should match snapshot for individual components: component-1-Components/MyOtherComponent 1`] = `
{
  "description": "",
  "id": "components/myothercomponent",
  "name": "Components/MyOtherComponent",
  "props": [],
  "stories": {
    "Components/MyOtherComponent": {
      "id": "components-myothercomponent--simple",
      "importPath": "./src/components/MyOtherComponent.stories.jsx",
      "kind": "Example/Button",
      "name": "Simple",
      "parameters": {
        "__id": "components-myothercomponent--simple",
        "docsOnly": false,
        "fileName": "./src/components/MyOtherComponent.stories.jsx",
      },
      "story": "Simple",
      "title": "Components/MyOtherComponent",
    },
  },
}
`;

exports[`storybook-api should match snapshot for individual components: component-1-Components/MyOtherComponent-props 1`] = `[]`;

exports[`storybook-api should match snapshot for individual components: component-1-Components/MyOtherComponent-story-Components/MyOtherComponent 1`] = `
{
  "id": "components-myothercomponent--simple",
  "importPath": "./src/components/MyOtherComponent.stories.jsx",
  "kind": "Example/Button",
  "name": "Simple",
  "parameters": {
    "__id": "components-myothercomponent--simple",
    "docsOnly": false,
    "fileName": "./src/components/MyOtherComponent.stories.jsx",
  },
  "story": "Simple",
  "title": "Components/MyOtherComponent",
}
`;
